src/openms/source/QC/DBSuitability.cpp

```diff
@@ -55,52 +65,387 @@ namespace OpenMS
+      OPENMS_LOG_ERROR << "Standard error: " << proc_stderr << endl;
+      throw Exception::InternalToolError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Return state was: " + static_cast<Int>(rt));
+    }
+    // search was successful
+
+    // load result
+    vector<ProteinIdentification> prot_ids;
+    vector<PeptideIdentification> pep_ids;
+    IdXMLFile id_file;
+    id_file.load(out_path, prot_ids, pep_ids);
+
+    if (keep_files)
+    {
+      id_file.store(tmp_dir.getPath() + "id_search_out.idXML", prot_ids, pep_ids);
+    }
+
+    // annotate target/decoy information
+    PeptideIndexing indexer;
+    FASTAContainer<TFI_Vector> proteins(fasta_data);
+    OPENMS_LOG_DEBUG << "Running PeptideIndexer functionalities ..." << endl << endl;
+    OPENMS_LOG_INFO.remove(cout); // prevent indexer from writing statistic
+    PeptideIndexing::ExitCodes indexer_exit = indexer.run(proteins, prot_ids, pep_ids);
+    OPENMS_LOG_INFO.insert(cout); // revert logging change
+    if (indexer_exit != PeptideIndexing::ExitCodes::EXECUTION_OK)
+    {
+      OPENMS_LOG_ERROR << "An error occured while trying to index the search results." << endl;
+      throw Exception::InternalToolError(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Return state was: " + static_cast<Int>(indexer_exit));
+    }
+
+    if (keep_files)
+    {
+      id_file.store(tmp_dir.getPath() + "indexed_pre_FDR.idXML", prot_ids, pep_ids);
+    }
+
+    // calculate q-values
+    FalseDiscoveryRate fdr;
+    Param p(fdr.getParameters());
+    if (!p.exists("use_all_hits") || !p.exists("add_decoy_peptides") || !p.exists("add_decoy_proteins"))
+    {
+      throw Exception::InvalidParameter(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "FDR parameters probably changed. 'use_all_hits', 'add_decoy_peptides' or 'add_decoy_proteins' not found.");
+    }
     p.setValue("use_all_hits", "true");
     p.setValue("add_decoy_peptides", "true");
     p.setValue("add_decoy_proteins", "true");
 
-    FalseDiscoveryRate fdr;
     fdr.setParameters(p);
+    OPENMS_LOG_DEBUG << "Calculating q-values ..." << endl << endl;
     fdr.apply(pep_ids);
 
+    if (keep_files)
+    {
+      id_file.store(tmp_dir.getPath() + "indexed_post_FDR.idXML", prot_ids, pep_ids);
+    }
+
+    return pep_ids;
+  }
+
+  std::vector<FASTAFile::FASTAEntry> DBSuitability::getSubsampledFasta_(std::vector<FASTAFile::FASTAEntry> fasta_data, double subsampling_rate) const
+  {
+    if (subsampling_rate < 0 || subsampling_rate > 1)
+    {
+      throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Subsampling rate has to be between 0 and 1. Aborting!");
+    }
+    Size num_AA{};
+    for (const auto& entry : fasta_data)
+    {
+      num_AA += entry.sequence.size();
+    }
+    double num_AS_written = num_AA * subsampling_rate;
+
+    //random_device rd();
+    //mt19937 g(rd());
+    mt19937 g(0);
+    shuffle(fasta_data.begin(), fasta_data.end(), g);
```

Comment:
How about shuffling the indices instead? Then you can take the fasta as const ref.